### ISSUE TITLE
DependencyCycle on second compile after 1-change compilation

### ISSUE THREAD

**Issue by da2x** [nanoc/nanoc#1619]
Regression in version 4.12.11.

``` 
nanoc compile
touch content/entry/something-thats-tagged-with-android.html
nanoc compile  # works just fine
nanoc compile  # this one fails with a DependencyCycle
nanoc compile  # this one fails too
```

```
Nanoc::Core::Errors::DependencyCycle: The site cannot be compiled because there is a dependency cycle:
                                                                                                                             
    (1) item /topic/android-1.html, rep :default, uses compiled content of                                                    
    (2) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (3) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (4) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (5) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (6) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (7) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (8) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (9) item /topic/sitemap.xml, rep :default, uses compiled content of                                                      
    (10) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (11) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (12) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (13) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (14) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (15) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (16) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (17) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (18) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (19) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (20) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (21) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (22) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (23) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (24) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (25) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (26) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (27) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (28) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (29) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (30) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (31) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (32) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (33) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (34) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (35) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (36) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (37) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (38) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (39) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (40) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (41) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (42) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (43) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (44) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (45) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (46) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (47) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (48) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (49) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (50) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (51) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (52) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (53) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (54) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (55) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (56) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (57) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (58) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (59) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (60) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (61) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (62) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (63) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (64) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (65) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (66) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (67) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (68) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (69) item /topic/sitemap.xml, rep :default, uses compiled content of                                                     
    (70) item /topic/sitemap.xml, rep :default, uses compiled content of (1)
```

work-around:

```
rm -rf tmp/nanoc/
nanoc compile
```

The sitemap template (below) does depend on the first URL. However, the first URL does not depend on the sitemap. It’s a one-way dependency. The first URL has no links to or attributes from the sitemap.

```
<?xml version="1.0" encoding="utf-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
<% @items.find_all('/topic/*.html').sort_by { |i| canonical_uri(i) }.each do |item| %>
  <url>
    <loc><%= canonical_uri(item) %></loc>
    <lastmod><%
       raw_path = item.reps[:default].raw_path
       sleep 1 unless File.exist?(raw_path)  # ugly
    %><%=
       File.mtime(raw_path).utc.strftime('%FT%R:00Z') %></lastmod>
    <changefreq>monthly</changefreq>
  </url>
<%  end %>
</urlset>
```

I rewrote my site to get the mtimes during precompile instead, but logging the issue as others may run into it.

**Comment by denisdefreyne** [comment]
Oh no! I’ve got a good hunch why this happens -- likely an oversight when I made https://github.com/nanoc/nanoc/pull/1609. Will work on a fix.

**Comment by denisdefreyne** [comment]
I believe I have a fix in https://github.com/nanoc/nanoc/pull/1620.

@da2x If you’ve got the old code still, could you give it a try and see whether it fixes it for you? I could reproduce it, but your site would be a great place to test it out too.

---

Two more things:

* `sleep 1 unless File.exist?(raw_path)  # ugly` -- you shouldn’t need this, as Nanoc does something like this internally already (even with the ugly wait!)

* Items that come from the filesystem datasource automatically get a `:mtime` attribute, based on the content filename (or meta filename, whichever is newer). Maybe this works for you as well?

**Comment by da2x** [comment]
> I believe I have a fix in https://github.com/nanoc/nanoc/pull/1620.

I can’t reproduce on that branch with my old code. I can’t immediately found anything else that broke either.

> `sleep 1 unless File.exist?(raw_path) # ugly` -- you shouldn’t need this, as Nanoc does something like this internally already (even with the ugly wait!)

This used to fail frequently because the task began before the dependency had been written to disk. The 1 second sleep reduced the failure rate from 1/25 times to twice a year.

> Items that come from the filesystem datasource automatically get a :mtime attribute, based on the content filename (or meta filename, whichever is newer). Maybe this works for you as well?

These items, a set of tag collection pages, doesn’t exist in the file system. They’re generated in the preprocessor by grabbing all the item[:tags] from a set of file-system backed items. So, I can’t rely on that for this.

**Comment by denisdefreyne** [comment]
A fix for this is released as part of Nanoc 4.12.12 (https://nanoc.app/release-notes/)
